src/lib.rs

```diff
@@ -0,0 +1,177 @@
+            api_key: api_key.to_string(),
+        }
+    }
+
+    fn url_for_endpoint(&self, endpoint: &str) -> String {
+        let mut url: String = self.base_url.to_string();
+        url.push_str(self.version);
+        url.push_str(endpoint);
+        url
+    }
+
+    fn add_headers(&self, builder: RequestBuilder) -> Result<RequestBuilder, ApiError> {
+        let mut headers = HeaderMap::new();
+        headers.insert("X-Pwinty-MerchantId", self.merchant_id.parse()?);
+        headers.insert("X-Pwinty-REST-API-Key", self.api_key.parse()?);
+        headers.insert("Content-Type", "application/json".parse()?);
+        headers.insert("accept", "application/json".parse()?);
+        Ok(builder.headers(headers))
+    }
+
+    fn make_request<Resp: 'static + DeserializeOwned + Send>(
+        &self,
+        endpoint: &str,
+        method: Method,
+        add_headers: bool,
+        body: Option<String>,
+    ) -> Box<dyn Future<Item = Resp, Error = ApiError> + Send> {
+        let url = self.url_for_endpoint(endpoint);
+        let client = if add_headers {
+            match self.add_headers(Client::new().request(method, &url)) {
+                Ok(res) => res,
+                Err(e) => {
+                    return Box::new(future::err(ApiError::InternalError(format!("{:?}", e))))
+                }
+            }
+        } else {
+            Client::new().request(method, &url)
+        };
+
+        let request = if let Some(request_body) = body {
+            client.body(request_body)
+        } else {
+            client
+        };
+        Box::new(
+            request
+                .send()
+                .map_err(|e| ApiError::RequestError(e))
+                .and_then(|mut res: Response| {
+                    if !res.status().is_success() {
+                        return future::err(ApiError::ResponseError {
+                            status: res.status(),
+                        });
+                    }
+                    future::ok(res.json::<Resp>().map_err(|e| e.into()))
+                })
+                .flatten(),
+        )
+    }
+
+    pub fn countries(
+        &self,
+    ) -> Box<dyn Future<Item = Vec<payloads::Country>, Error = ApiError> + Send> {
+        Box::new(
+            self.make_request::<payloads::Countries>("/countries", Method::GET, false, None)
+                .and_then(|countries| Ok(countries.data)),
+        )
+    }
+
+    pub fn create_order(
+        &self,
+        order_info: &payloads::OrderCreate,
+    ) -> Box<dyn Future<Item = payloads::Order, Error = ApiError> + Send> {
+        let body = match serde_json::to_string(&order_info) {
+            Ok(body) => body,
+            Err(e) => return Box::new(future::err(ApiError::InternalError(format!("{:?}", e)))),
+        };
+
+        Box::new(
+            self.make_request::<payloads::OrderWrapper>("/orders", Method::POST, true, Some(body))
+                .and_then(|order| future::ok(order.data)),
```

Comment:
Thoughts on `future::ok()` vs `Ok()`?

Reply:
I don't know tokio / async await well enough to say one way or the other haha